### PR TITLE
add gpio driver sn74hc595

### PR DESCRIPTION
**Pull request by bimokn97**
## Description
Add sn74595 serial output chip driver using SPI.
recreate and little change from previous PR

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/gpio/sn74hc595.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    spi_start(SN74HC595_SPI_SLAVE_SELECT_PIN, SN74HC595_SPI_LSBFIRST, SN74HC595_SPI_MODE, SN74HC595_SPI_CLOCK_DIVISOR);
+    writePinLow(SN74HC595_SPI_SLAVE_SELECT_PIN);
+    spi_transmit(pinStates, SN74HC595_LENGTH);
+    writePinHigh(SN74HC595_SPI_SLAVE_SELECT_PIN);
+    spi_stop();
+
+// pinstates debug
+#if defined(CONSOLE_ENABLE) && defined(SN74HC595_DEBUG)
+    if (timer_elapsed32(timer_now) >= 1000) {
+        uprintf("PinStates HEX: ");
```
Comment (drashna) [inline drivers/gpio/sn74hc595.c:64]: if you use dprintf here, it only prints if debug_enable is set to true. which can be toggled at runtime.

**Review by drashna — APPROVED** [review]

File: drivers/gpio/sn74hc595.c (hunk 1 above)
Comment (bimokn97) [reply drivers/gpio/sn74hc595.c:64]: i will fix it on next, also will fix other line on next.

File: drivers/gpio/sn74hc595.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    spi_stop();
+    return res;
+}
+
+/*
+    set pin number on sn74hc595
+*/
+spi_status_t sn74hc595_setPin(pin_t pin, bool set) {
```
Comment (zvecr) [inline drivers/gpio/sn74hc595.c:59]: `pin_t` should not be used for this purpose, its use is limited to mcu pins.

**Review by zvecr — CHANGES_REQUESTED** [review]
The introduced API seems a little inconsistent to existing GPIO drivers. For example, exposing transport errors as return types.

**Comment by bimokn97** [comment]
> The introduced API seems a little inconsistent to existing GPIO drivers. For example, exposing transport errors as return types.

The API it self don't have return value, So is it ok just use void function(){} with no return?,
sorry for my worst code, need advice.

**Pull request closed without merging**